includes/API/Packs/ApiLabkiPacksList.php

```diff
@@ -0,0 +1,650 @@
+ * - Uses `LabkiPageRegistry` for page data
+ * - Uses `LabkiRepoRegistry` and `LabkiRefRegistry` for resolution
+ * - Supports filtering by repo, ref, and pack
+ * - Supports both ID and name lookups for repo, ref, and pack
+ * - Returns empty array if no packs found
+ *
+ * @ingroup API
+ */
+class ApiLabkiPacksList extends PackApiBase {
+
+	/**
+	 * Constructor.
+	 *
+	 * @param \ApiMain $main Main API object
+	 * @param string $name Module name
+	 */
+	public function __construct( \ApiMain $main, string $name ) {
+		wfDebugLog( 'labkipack', "ApiLabkiPacksList::__construct() called with name={$name}" );
+		parent::__construct( $main, $name );
+	}
+
+	/**
+	 * Execute the API request.
+	 *
+	 * Main entry point for the API. Handles:
+	 * - Listing all packs (no parameters)
+	 * - Listing packs for a repository (repo_id or repo_url)
+	 * - Listing packs for a specific ref (repo + ref or ref_id)
+	 * - Listing a specific pack (repo + ref + pack or pack_id)
+	 *
+	 * All modes return consistent structure: { packs: [...] }
+	 * Use include_pages=true to nest page data within each pack.
+	 */
+	public function execute(): void {
+		wfDebugLog( 'labkipack', "ApiLabkiPacksList::execute() started" );
+
+		// Get parameters
+		$repoId = $this->getParameter( 'repo_id' );
+		$repoUrl = $this->getParameter( 'repo_url' );
+		$refId = $this->getParameter( 'ref_id' );
+		$ref = $this->getParameter( 'ref' );
+		$packId = $this->getParameter( 'pack_id' );
+		$pack = $this->getParameter( 'pack' );
+		$includePages = $this->getParameter( 'include_pages' );
+
+		wfDebugLog( 'labkipack', "ApiLabkiPacksList::execute() params: repo_id={$repoId}, repo_url={$repoUrl}, ref_id={$refId}, ref={$ref}, pack_id={$packId}, pack={$pack}, include_pages={$includePages}" );
+
+		// Validate: only one repo identifier should be provided
+		if ( $repoId !== null && $repoUrl !== null ) {
+			$this->dieWithError(
+				'labkipackmanager-error-repo-multiple-identifiers',
+				'multiple_identifiers'
+			);
+		}
+
+		// Validate: only one ref identifier should be provided
+		if ( $refId !== null && $ref !== null ) {
+			$this->dieWithError(
+				'labkipackmanager-error-ref-multiple-identifiers',
+				'multiple_identifiers'
+			);
+		}
+
+		// Validate: only one pack identifier should be provided
+		if ( $packId !== null && $pack !== null ) {
+			$this->dieWithError(
+				'labkipackmanager-error-pack-multiple-identifiers',
+				'multiple_identifiers'
+			);
+		}
+
+		// Validate: if ref is specified, repo must be specified
+		if ( ( $refId !== null || $ref !== null ) && $repoId === null && $repoUrl === null ) {
+			$this->dieWithError(
+				'labkipackmanager-error-ref-requires-repo',
+				'missing_repo'
+			);
+		}
+
+		// Validate: if pack is specified, ref must be specified (unless using pack_id)
+		if ( $pack !== null && $refId === null && $ref === null ) {
+			$this->dieWithError(
+				'labkipackmanager-error-pack-requires-ref',
+				'missing_ref'
+			);
+		}
```

Comment:
if this is a recurring calling pattern, definitely wrap it up in a method so we don't need a hundred lines of arg parsing in every endpoint